File: Episode08-Explosions/Love2D/explosion.lua
local Class = require("lib.Class")
local Explosion = Class:derive("Explosion")
local Rectangle = require "lib.Rectangle"

function Explosion:new(spriteList, centre, scale)
	--[[ class constructor, takes table of animation images
		 centre of the mob destroyed {x,y} , and scale factor]]
	self.centre = centre
	self.spriteList = spriteList 	-- table of images
	self.scale = scale
	self.image = spriteList[1] 		-- set to first image in the sequence
	self.frame = 1
	self.timePassed = 0
	self.frameRate = 0.1
	self.active = true
	self.rect = Rectangle(centre[1] - (self.image:getWidth() / 2) * self.scale,
						  centre[2] - (self.image:getHeight() / 2) * self.scale,
						  self.image:getWidth(),
						  self.image:getHeight())
end

function Explosion:update(dt)
	self.timePassed = self.timePassed + dt
	if self.timePassed >= self.frameRate then
		self.timePassed = 0
		self.frame = self.frame + 1
		if self.frame > #self.spriteList then
			self.frame = #self.spriteList
			self.active = false
		else
			-- ensure image sequence is centered per frame 
			self.image = self.spriteList[self.frame] --next frame
			self.rect:update(self.centre[1] - (self.image:getWidth() / 2) * self.scale,
							 self.centre[2] - (self.image:getHeight() / 2) * self.scale)
		end
	end
	return self.active -- when false, explosion can be set to nil
end

function Explosion:draw()
	-- love.graphics.draw( drawable, x, y, r, sx, sy, ox, oy, kx, ky )
	if self.active then
		love.graphics.draw(self.image, self.rect.x, self.rect.y, nil, self.scale, self.scale)
	end
end

return Explosion

File: Episode08-Explosions/Pygame/explosion.py
import pygame
import shared

class Explosion():
	def __init__(self, images, center, key):
		self.key = key
		self.images = images
		self.image = images[self.key][0] #set to first image in the sequence
		self.rect = self.image.get_rect()
		self.rect.center = center
		self.frame = 0
		self.time_passed = 0
		self.frame_rate = 0.1 
		self.active = True

	def update(self, dt):
		self.time_passed += dt
		if self.time_passed >= self.frame_rate:
			self.time_passed = 0
			self.frame += 1
			if self.frame >= len(self.images[self.key]):
				self.active = False
			else:
				# ensure image sequence is centered per frame
				center = self.rect.center 
				self.image = self.images[self.key][self.frame] #next frame
				self.rect = self.image.get_rect()
				self.rect.center = center
				
		return self.active
	
	def draw(self):
		shared.screen.blit(self.image, self.rect)